LUCENE-10054 Handle hierarchy in graph construction and search

mayya-sharipova opened the pull request:
This patch handles hierarchy in graph construction and search,
but only in memory.

Changes:

- `HnswGraphBuilder` has an extra parameter :`ml` normalization factor for level generation.  When `ml=0`,  the graph will have only a single layer: SNW. When `ml > 0`, the graph will have multiple layers.  
     - The recommended `ml` value from 2018 HNSW paper is : `ml = 1 / Math.log(1.0 * maxConn)`, which was used for tests in `TestHnswGraph` class.
    - When `ml = 0`,  we use the previous code in the method `buildSNW`
    - When `ml > 0`,  the method `buildHNSW` is used, that according the paper, for every new node: generates a random level for this node, and then places the new node in those levels with its connections.
- `HnswGraph`'s `search` method has also been modified to handle two cases:
   - When `ml = 0`,  we use the previous code for a flat graph: generate `boundedNumSeed` number of random entry points, and use them to search a flat graph
   - When `ml > 0`, we use the hierarchical graph: from max level till level 1 using `topK=1`, and on the level 0th using `topK=boundedNumSeed`


Work left for future: handle hierarchy on disk

mayya-sharipova committed:
LUCENE-10054 Handle graph hierarchy in memory

This patch handles hierarchy in graph construction and search,
but only in memory.

Changes:

- HnswGraphBuilder has an extra parameter :ml normalization factor for level
  generation. When ml=0, the graph will have only a single layer: SNW.
  When ml > 0, the graph will have multiple layers.
  - The recommended ml value from 2018 HNSW paper is :
    ml = 1 / Math.log(1.0 * maxConn), which was used for tests in
    TestHnswGraph class.
  - When ml = 0, we use the previous code in the method buildSNW
  - When ml > 0, the method buildHNSW is used, that according the paper,
    for every new node: generates a random level for this node, and then
     places the new node in those levels with its connections.
- HnswGraph's search method has also been modified to handle two cases:
  - When ml = 0, we use the previous code for a flat graph: generate
    boundedNumSeed number of random entry points, and use them to search
    a flat graph
  - When ml > 0, we use the hierarchical graph: from max level till level 1
    using topK=1, and on the level 0th using topK=boundedNumSeed

Work left for future: handle hierarchy on disk

jtibshirani commented on lucene/core/src/java/org/apache/lucene/util/hnsw/HnswGraph.java:
```diff
@@ -206,11 +275,65 @@ public void seek(int level, int targetNode) {
+   *
+   * <p>Package private access to use only for tests
+   */
+  void seekLevel(int level) {
```
It feels a little confusing that we have both `seekLevel(level)`, which positions the level, and `seek(level, targetNode)` which doesn't. Maybe we can refine this API once as part of the on-disk PR, when we'll have a better idea of the final graph layout.

mayya-sharipova replied:
Great comment. It is indeed confusing.
I've redesigned KnnGraphValues methods to have: 
- A single `seekLevel(int level)` to position on a given level
- `seek(targetNode)` to position on an given node of the previously positioned level

jtibshirani reviewed (commented):
I left a few comments but haven't caught up up on all the details yet. A general comment: I wonder if we should really allow `ml` to be configurable. I think other HNSW implementations choose this using a heuristic and don't allow it to be set. This would let us drop the alternate code paths for graphs with only one level and remove old naming like `numSeed`.

From playing around locally, it also looks like `TestHnswGraph` has a test failure. Example:
```
./gradlew test --tests TestHnswGraph.testSearchWithAcceptOrds -Dtests.seed=E7C78E0C128E2469
```

msokolov commented on lucene/core/src/java/org/apache/lucene/util/hnsw/HnswGraph.java:
```diff
@@ -188,15 +244,28 @@ public int size() {
+        curMaxLevel = level;
+        entryNode = node;
+        // add more levels if needed
+        if (level >= graph.size()) {
```
Wait - what does `graph.size()` mean here? Is it the number of nodes in level 0? Or the number of levels? Oh I remember - graph is simply a list of levels. I wonder if it would be clearer to call it `graphLevels`?

mayya-sharipova replied:
I've modified the code to use `curMaxLevel` instead of `graph.size()`.

msokolov commented on lucene/core/src/java/org/apache/lucene/util/hnsw/HnswGraph.java:
```diff
@@ -188,15 +244,28 @@ public int size() {
+          }
+        }
+      }
       // Levels above 0th don't contain all nodes,
```
to be more minimalistic here, I think we need three pointers (ints) - one within this level, one to the next level, and one to level 0. Not every level needs them all; eg for level 0 and 1 some of them are redundant. If we use the current data structure to store the within-level pointers and add two (possibly null) int[] arrays for the other pointers, I think it will be nice and compact. Fine to save it for later - just thinking out loud how this might end up looking

mayya-sharipova replied:
@msokolov  Are you suggesting this for the whole graph organization?  Instead of having a list of levels?

msokolov commented on lucene/core/src/java/org/apache/lucene/util/hnsw/HnswGraphBuilder.java:
```diff
@@ -146,20 +168,72 @@ void addGraphNode(int node, float[] value) throws IOException {
+    // start at node 1! node 0 is added implicitly, in the constructor
+    for (int node = 1; node < vectors.size(); node++) {
+      addGraphNodeHNSW(node, vectors.vectorValue(node));
+      if (node % 10000 == 0) {
```
can we refactor and share with the other place we do this?

mayya-sharipova replied:
Addressed in 6a05951772cc72a1530f3fd863d906dc0e3bef88

msokolov commented on lucene/core/src/java/org/apache/lucene/util/hnsw/HnswGraphBuilder.java:
```diff
@@ -267,4 +341,12 @@ private int findNonDiverse(NeighborArray neighbors) throws IOException {
+
+  private static int getRandomGraphLevel(double ml, Random random) {
+    float randFloat = random.nextFloat();
+    while (randFloat == 0.0f) {
```
looks like a do-while to me!

mayya-sharipova replied:
good comment! Addressed in 6a05951772cc72a1530f3fd863d906dc0e3bef88

msokolov commented on lucene/core/src/test/org/apache/lucene/util/hnsw/TestHNSWGraph2.java:
```diff
@@ -0,0 +1,162 @@
+      } else {
+        assertFalse("Graph has orphan nodes with no friends on level [" + level + "]", foundOrphan);
+        if (maxConn > nodesCount) {
+          // assert that the graph in fully connected, i.e. any node can be reached from any other
```
"is fully"

mayya-sharipova replied:
Addressed in 6a05951772cc72a1530f3fd863d906dc0e3bef88

msokolov reviewed (commented):
nice!

mayya-sharipova committed:
Address Julie & Mike's feedback

mayya-sharipova commented on lucene/core/src/java/org/apache/lucene/util/hnsw/HnswGraph.java:
```diff
@@ -154,7 +210,7 @@ public static NeighborQueue search(
-        if (results.size() < numSeed || bound.check(score) == false) {
+        if (results.size() < topK || bound.check(score) == false) {
           candidates.add(friendOrd, score);
           if (acceptOrds == null || acceptOrds.get(friendOrd)) {
```
Great comment! Addressed in 6a05951772cc72a1530f3fd863d906dc0e3bef88

mayya-sharipova commented:
@jtibshirani  Thanks for your review. I've addressed in 6a05951772cc72a1530f3fd863d906dc0e3bef88


> I wonder if we should really allow ml to be configurable. I think other HNSW implementations choose this using a heuristic and don't allow it to be set. This would let us drop the alternate code paths for graphs with only one level and remove old naming like numSeed.

I took an inspiration from the `nmslib` library where this parameter [is configurable](https://github.com/nmslib/nmslib/blob/master/similarity_search/src/method/hnsw.cc#L203).   Also it allows us to have a flat graph (`sw-graph` from `nmslib`) for cases where we want it.  Not sure though if we still want to keep flat graph, happy to continue the discussion.

---


> ./gradlew test --tests TestHnswGraph.testSearchWithAcceptOrds -Dtests.seed=E7C78E0C128E2469


This test fails because it ends up marking many expected results docs (from 0 to 9) as deleted which don't get retrieved.
I've modified `createRandomAcceptOrds(int startIndex, int length)` to add `startIndex` from which we want to start marking documents deleted.

The reason this test was not failing on the main branch is because there we use `numSeed=5` and return 5 results instead of 5 and their sum always ends up < 75.

mayya-sharipova commented on lucene/core/src/java/org/apache/lucene/util/hnsw/HnswGraph.java:
```diff
@@ -107,32 +113,82 @@ public static NeighborQueue search(
+      }
+      return searchLevel(query, topK, 0, eps, vectors, similarityFunction, graphValues, acceptOrds);
+    } else {
+      // search in hierarchical SNW
```
Thanks, good comment. Addressed in 6a05951772cc72a1530f3fd863d906dc0e3bef88

mayya-sharipova commented on lucene/core/src/java/org/apache/lucene/util/hnsw/HnswGraph.java:
```diff
@@ -107,32 +113,82 @@ public static NeighborQueue search(
-        if (acceptOrds == null || acceptOrds.get(entryPoint)) {
-          results.add(entryPoint, score);
+
+    for (int i = 0; i < eps.length; i++) {
```
Good comment, addressed in 6a05951772cc72a1530f3fd863d906dc0e3bef88

mayya-sharipova commented on lucene/core/src/java/org/apache/lucene/util/hnsw/HnswGraph.java:
```diff
@@ -107,32 +113,82 @@ public static NeighborQueue search(
+        results =
+            HnswGraph.searchLevel(
+                query, 1, level, eps, vectors, similarityFunction, graphValues, acceptOrds);
+        eps = new int[] {results.pop()};
```
Good comment, addressed in 6a05951

mayya-sharipova commented on lucene/core/src/java/org/apache/lucene/util/hnsw/HnswGraph.java:
```diff
@@ -107,32 +113,82 @@ public static NeighborQueue search(
+      // search in SNW; generate a number of entry points randomly
+      final int[] eps = new int[boundedNumSeed];
+      for (int i = 0; i < boundedNumSeed; i++) {
+        eps[i] = random.nextInt(size);
```
I think the resulted behaviour here is the same as in the current implementation on the main branch.  Even though we may get repeats here,  `searchLevel()`  that follows, will discard these repeats with  a condition `if (visited.get(ep) == false) `

mayya-sharipova commented:
@msokolov Thanks very much for your feedback. I've tried to address  in 6a05951772cc72a1530f3fd863d906dc0e3bef88

jtibshirani commented:
> I took an inspiration from the `nmslib` library where this parameter [is configurable](https://github.com/nmslib/nmslib/blob/master/similarity_search/src/method/hnsw.cc#L203)

Thanks, I didn't poke through the C++ and had missed that. It still might not be common to set it, it looks like the HNSW implementations in ann-benchmarks don't tune over `ml`. I don't feel strongly, but we usually try to minimize configuration options, only exposing ones that are really useful and we expect users to tune. Maybe we could at least prevent `ml` from being `0` so we could remove the flat graph code path? This would be a nice simplification and I can't think of a use case for keeping a flat graph implementation as part of the HNSW codec.

> This test fails because it marks many expected results docs (from 0 to 9) as deleted which don't end up being retrieved.

That makes sense. This feels like an independent fix, maybe we could pull it into its own PR against `main` so it doesn't just sit on the feature branch?

mayya-sharipova commented:
@msokolov Thanks for explanation. Your idea is clear to me and makes sense, and I will implement it in the following PR.  I was just thinking if having only  `level0Node[]` is enough. I see several benefits with it:
- no need for an extra array  `nextLevelNode[]` and no need to keep it updated
- `searchLevel` consistently returns nodes ids of 0th level

Extra work: need to do binary search to map node0 ids to the current level's node ids for search to start.

WDYT?

msokolov commented:
> Extra work: need to do binary search to map node0 ids to the current level's node ids for search to start. Should not be that bad, as we use eps.length = 1 for levels > 0.

Agreed, we don't need `nextLevelNode`, this extra work is going to be negligible

mayya-sharipova commented:
@jtibshirani 

> Maybe we could at least prevent ml from being 0 so we could remove the flat graph code path? This would be a nice simplification and I can't think of a use case for keeping a flat graph implementation as part of the HNSW codec.

I agree, this indeed will be a good simplification. Just wanted to confirm that we are ok with not having a flat graph implementation? I've checked [ann-benchmarks](https://github.com/erikbern/ann-benchmarks), and on their datasets HNSW always outperforms SW-graph, so I think it is ok to have HNSW only.

jtibshirani commented on lucene/core/src/java/org/apache/lucene/util/hnsw/HnswGraph.java:
```diff
@@ -188,29 +244,95 @@ public int size() {
   }
 
   // TODO: optimize RAM usage so not to store references for all nodes for levels > 0
+  // TODO: add extra levels if level >= numLevels
```
Do we still need this TODO?

mayya-sharipova replied:
Removed in 7c82c8aa0d13d9202d40feb8ea71c6f23000c6fd

jtibshirani commented on lucene/core/src/java/org/apache/lucene/util/hnsw/HnswGraphBuilder.java:
```diff
@@ -66,13 +69,15 @@
       int beamWidth,
-      long seed) {
+      long seed,
+      double ml) {
```
Super small comment, this could go above seed to be grouped with other parameters `maxConn` and `beamWidth`.

mayya-sharipova replied:
Changed the place for this parameter in 7c82c8aa0d13d9202d40feb8ea71c6f23000c6fd

jtibshirani commented on lucene/core/src/java/org/apache/lucene/codecs/lucene90/Lucene90HnswVectorsReader.java:
```diff
@@ -480,8 +480,12 @@ private void readValue(int targetOrd) throws IOException {
       this.dataIn = dataIn;
     }
 
+    // TODO : implement hierarchy
```
Do we still need this TODO?

mayya-sharipova replied:
Removed in 7c82c8aa0d13d9202d40feb8ea71c6f23000c6fd

jtibshirani reviewed (approved):
> Can we remove the flat graph part when we implement on the disk storage? 

This works for me.

The change makes sense to me as an incremental step, I only left some minor comments. I'll likely have more thoughts on the naming/ implementation details in the next PRs that add the on-disk structure.

mayya-sharipova committed:
Address Julie's feedback 2
